### ISSUE TITLE
Replace MSBuild targets for xplat build of xproj/project.json new project format

### ISSUE THREAD

**Issue by galvesribeiro** [automotiveMastermind/condo#2]
Hello,

As part of Project Orleans move to CoreCLR, we are first porting the current *.csproj to the new .xproj, so we can target both dnx451 and dnxcore50 on the same project build.

The porting progress is tracked by this issue: https://github.com/dotnet/orleans/issues/368

Right now, we have on one of the projects, some MSBuild tasks that we need to port to the .xproj as follow:

```
<Target Name="OrleansDllBootstrapUsingCodeGen" AfterTargets="BeforeCompile" BeforeTargets="CoreCompile" Inputs="@(Compile);@(ReferencePath)" Outputs="$(ProjectDir)$(IntermediateOutputPath)Generated\$(TargetName)$(TargetExt)" Condition="'$(Bootstrap)' != 'true'">
    <Message Text="[OrleansDllBootstrapUsingCodeGen] - Bootstrapping the Orleans.dll" Importance="high" />
    <!-- Clean Generated directory -->
    <Message Text="[OrleansDllBootstrapUsingCodeGen] - Deleting/cleaning generated files for Project=$(ProjectName)" />
    <MakeDir Directories="$(IntermediateOutputPath)Generated" />
    <Touch Files="$(ProjectDir)Properties\orleans.codegen.cs" ForceTouch="true" AlwaysCreate="true" ContinueOnError="true" Condition="!Exists('$(ProjectDir)Properties\orleans.codegen.cs')" />
    <PropertyGroup>
      <BootstrapOutputPath>$(OutDir)Bootstrap\</BootstrapOutputPath>
      <BootstrapOutputPath Condition="'$(BuildingInsideVisualStudio)' == 'true'">$(ProjectDir)$(OutDir)Bootstrap\</BootstrapOutputPath>
      <ExcludeCodeGen>$(DefineConstants);EXCLUDE_CODEGEN</ExcludeCodeGen>
    </PropertyGroup>
    <Message Text="[OrleansDllBootstrapUsingCodeGen] - OutputPath: $(BootstrapOutputPath)" Importance="high" />
    <!-- Compile code generator -->
    <Message Text="[OrleansDllBootstrapUsingCodeGen] - Compiling Orleans.dll for bootstrap" Importance="high" />
    <MSBuild Projects="$(MSBuildProjectFullPath)" Targets="Build" Properties="Bootstrap=true;BootstrapOutputPath=$(BootstrapOutputPath);DefineConstants=$(ExcludeCodeGen)" UnloadProjectsOnCompletion="true" UseResultsCache="false" />
    <Message Text="[OrleansDllBootstrapUsingCodeGen] - Compiling code generators for bootstrap" Importance="high" />
    <MSBuild Projects="$(SolutionDir)\OrleansCodeGenerator\OrleansCodeGenerator.csproj" Targets="Build" Properties="Bootstrap=true;BootstrapOutputPath=$(BootstrapOutputPath);OutputPath=$(BootstrapOutputPath);OutDir=$(BootstrapOutputPath);DefineConstants=$(ExcludeCodeGen)" UnloadProjectsOnCompletion="true" UseResultsCache="false" />
    <MSBuild Projects="$(SolutionDir)\ClientGenerator\ClientGenerator.csproj" Targets="Build" Properties="Bootstrap=true;BootstrapOutputPath=$(BootstrapOutputPath);OutputPath=$(BootstrapOutputPath);OutDir=$(BootstrapOutputPath);DefineConstants=$(ExcludeCodeGen)" UnloadProjectsOnCompletion="true" UseResultsCache="false" />
    <!-- Finally invoke code generator on the recently built Orleans.dll -->
    <Message Text="[OrleansDllBootstrapUsingCodeGen] - Preprocessing $(TargetName)$(TargetExt) with ClientGenerator.exe" Importance="high" />
    <PropertyGroup>
      <ArgsFile>$(IntermediateOutputPath)$(TargetName).codegen.args.txt</ArgsFile>
    </PropertyGroup>
    <ItemGroup>
      <CodeGenArgs Include="/bootstrap" />
      <CodeGenArgs Include="/in:$(BootstrapOutputPath)$(TargetName)$(TargetExt)" />
      <CodeGenArgs Include="@(ReferencePath->'/r:%(Identity)')" />
      <CodeGenArgs Include="@(Compile->'/src:%(Identity)')" />
    </ItemGroup>
    <WriteLinesToFile Overwrite="true" File="$(ArgsFile)" Lines="@(CodeGenArgs)" />
    <Message Text="[OrleansDllBootstrapUsingCodeGen] - Invoking Code Generator" />
    <Exec Command="&quot;$(BootstrapOutputPath)ClientGenerator.exe&quot; &quot;@$(ArgsFile)&quot;" />
  </Target>
```

The idea there is to build one of the projects, a console app, and than use it to code-generation of the Orleans.csproj.

The big challenge here is that as we are going to be coreCLR compliant, the build must be supported on Linux/OSX (including using VS Code for that) so we can't rely on powershell scripts which would be a good replacement for that. We could have a build.ps1 and a build.sh for each environment if the only usage of the code generator was at the CI server however, we need it to run when building from Visual Studio/Code as well and that is why the MSBuild tasks were chosen at first place.

The other problem is that we can't install anything on the CI/build machine widely since we are using the .Net Foundation CI systems. So if something must be downloaded and run in order to perform the build, it must be done locally.

So, does Condo help us with that cases?

Thanks!

**Comment by dmccaffery** [comment]
I will look at this more deeply, but now that I am seeing what you're trying to accomplish, my initial thoughts are as follows:

Don't rely on MSBuild for anything, other than a basic wrapper around project.json (which is what an .xproj is) -- the only reason that XProj still exists is that the Visual Studio project management system is built around MSBuild, and it was easier to 'proxy' the DNX Host through an MSBuild task than it was to make Visual Studios project system project.json aware.

Obviously, you will still need to support your needs, in terms of executing some tasks before/after a build using both Visual Studio and the CLI. It turns out that this is actually fairly easy to accomplish, albeit a bit difficult to make cross-platform. The project system for DNX has support for executing scripts at different times within a build lifecycle: [Scripts in project.json](https://github.com/aspnet/Home/wiki/Project.json-file#scripts)

The difficulty here is, there is currently no way to change which script is executed based on the environment (Windows vs OS X vs Linux). The best way that I have been able to tackle this problem is through a task runner, such as gulp or grunt, which uses node.js underneath. Node.js already has a great cross-platform abstraction over the OS, such as accessing the file-system, etc. Since this introduces a crap load of dependencies that you wouldn't need for any other purpose, this isn't the right way to go in your case.

The "correct" way to do this would be to create a custom command for dnx that is published to NuGet or myGet out of band: see [DNX Commands](http://docs.asp.net/en/latest/dnx/commands.html#building-and-publishing-global-command): that performs the code generation (something like orleans-gen).

Then you can add a prebuild/postbuild script that calls this command (which will work xplat):

``` json
"scripts": {
  "prebuild": "dnx orleans-gen [optional arguments]"
}
```
Now, from this point, any time that dnx build is called against the project (which Visual Studio does when you build an xproj), that prebuild or postbuild script will also be called, which executes your generator. Since condo calls dnx build on your behalf as part of a CI build, it would also execute that command.

TL;DR:
* It's possible
* MSBuild is probably the most difficult way to do it
* Condo will help with CI if you want a fully-featured build system without having to write a lengthy build script of your own
* Use dnx commands and dnx scripts to execute tasks at the right point in the pipeline

I'll look at it a bit deeper, and maybe create a PR against Orleans or a gist to provide an example. I hope this is helpful. :)

**Comment by dmccaffery** [comment]
Condo now fully supports MSBuild integration through the installation of the Condo.MSBuild nuget package against each csproj. We currently only support csproj at the present time.

**Comment by dmccaffery** [comment]
Reopening to track request to execute out-of-band tool as a post-build step. While condo already supports this, it is poorly documented. Will update documentation and provide direct support to get this accomplished. :)
